ISSUE-1571 Add a fluent Instant assertion for OffsetDateTime

raynicho opened the pull request:
#### Check List:
* Fixes #1571 
* Unit tests : YES
* Javadoc with a code example (API only) : YES

Added a fluent `Instant` assertion for `OffsetDateTime` objects.

joel-costigliola commented:
Thanks, I will review your PR shortly !

joel-costigliola commented on src/main/java/org/assertj/core/api/AbstractOffsetDateTimeAssert.java:
```diff
@@ -671,6 +671,33 @@ public SELF isStrictlyBetween(String startExclusive, String endExclusive) {
+   * @throws IllegalArgumentException if other {@code OffsetDateTime} is {@code null}.
+   * @throws AssertionError if the actual {@code OffsetDateTime} does not have the same {@code Instant} as the other.
+   */
+  public SELF hasSameInstantAs(OffsetDateTime other) {
```
After giving it more time, I think I got a better name: `isAtSameInstantAs`, the idea came after reading [OffsetDateTime.html#withOffsetSameInstant](https://docs.oracle.com/javase/8/docs/api/java/time/OffsetDateTime.html#withOffsetSameInstant-java.time.ZoneOffset-) which states:
> Returns a copy of this OffsetDateTime with the specified offset ensuring that the result **is at the same instant**.

raynicho replied:
Got the method changed to `isAtSameInstantAs`.

joel-costigliola commented on src/test/java/org/assertj/core/api/offsetdatetime/OffsetDateTimeAssert_hasSameInstantAs_Test.java:
```diff
@@ -0,0 +1,61 @@
+  private final OffsetDateTime actual = of(2000, 12, 12, 3, 0, 0, 0, ZoneOffset.ofHours(3));
+
+  @Test
+  public void should_pass_if_has_same_instant() {
```
rename to `should_pass_if_at_the_same_instant`

raynicho replied:
Changed the test name to `should_pass_if_at_the_same_instant`.

joel-costigliola commented on src/test/java/org/assertj/core/api/offsetdatetime/OffsetDateTimeAssert_hasSameInstantAs_Test.java:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  @Test
+  public void should_fail_if_has_different_instant() {
```
rename to `should_fail_if_at_a_different_instant`

raynicho replied:
Changed the test case name to `should_fail_if_at_a_different_instant`.

joel-costigliola commented on src/main/java/org/assertj/core/api/AbstractOffsetDateTimeAssert.java:
```diff
@@ -671,6 +671,33 @@ public SELF isStrictlyBetween(String startExclusive, String endExclusive) {
+  public SELF hasSameInstantAs(OffsetDateTime other) {
+    Objects.instance().assertNotNull(info, actual);
+    assertOffsetDateTimeParameterIsNotNull(other);
+    comparables.assertEqual(info, actual.toInstant(), other.toInstant());
```
It would be better to introduce a specific error message stating we were comparing the OffsetDateTime as Instant.
This means adding a `ShouldBeAtSameInstant` with an error message like:
```
"%nExpecting%n" +
"  <%s>%n" +
"to be at the same instant as:%n" +
"  <%s>%n" +
"but actual instance was%n" +
"  <%s>%n" +
"and expected instant was:%n" +
"  <%s>"
```
A `ShouldBeAtSameInstant_create_Test` should be added.

You cab refer to `ShouldHaveSameHashCode` and `ShouldHaveSameHashCode_create_Test` for how to do this

joel-costigliola replied:
This also means `comparables.assertEqual` can't be used anymore, but something like:
```
if (!comparables.toInstant().equals(other.toInstant())) throw failures.failure(info, shouldBeAtSameInstant(actual, other));
```

raynicho replied:
Added a `ShouldBeAtSameInstant`, using it as the failure message for `isAtSameInstantAs`.

joel-costigliola commented on src/test/java/org/assertj/core/api/offsetdatetime/OffsetDateTimeAssert_hasSameInstantAs_Test.java:
```diff
@@ -0,0 +1,61 @@
+import org.assertj.core.api.BaseTest;
+import org.junit.jupiter.api.Test;
+
+public class OffsetDateTimeAssert_hasSameInstantAs_Test extends BaseTest {
```
nothing wrong with this class, it would be better to follow `ObjectArrays_assertHasSameSizeAs_with_Iterable_Test` ways, we found it to be a bit more readable ;-)

raynicho replied:
I changed `OffsetDateTimeAssert_isAtSameInstantAs_Test` to extend from `OffsetDateTimeAssertBaseTest`, adding two new static `OffsetDateTimes`. Let me know if I should have done something differently.

joel-costigliola commented:
@raynicho I have finished reviewing the PR, good work. Some changes are needed but that's not something you could have known beforehand.

raynicho commented:
@joel-costigliola Addressed your comments, let me know if I missed anything/got something wrong. I didn't want to squash and force push because git hasn't liked that in the past...

> When pushing commits to a pull request, don't force push. Force pushing can corrupt your pull request.

joel-costigliola commented:
Integrated thanks @raynicho!
